File: 3302-find-the-lexicographically-smallest-valid-sequence/3302-find-the-lexicographically-smallest-valid-sequence.c
#include <stdlib.h>
#include <string.h>

/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int* validSequence(char* word1, char* word2, int* returnSize) {
    int n = strlen(word1);
    int m = strlen(word2);

    // Highly efficient stack allocation or direct pointer calculations
    int* lastValidW1Idx = (int*)malloc(m * sizeof(int));
    for (int i = 0; i < m; i++) {
        lastValidW1Idx[i] = -1;
    }

    int j = m - 1;
    for (int i = n - 1; i >= 0; i--) {
        if (j >= 0 && word1[i] == word2[j]) {
            lastValidW1Idx[j] = i;
            j--;
        }
    }

    int* answer = (int*)malloc(m * sizeof(int));
    int w2Idx = 0;
    int changed = 0; // 0 represents false, 1 represents true

    for (int i = 0; i < n && w2Idx < m; i++) {
        char c1 = word1[i];
        char c2 = word2[w2Idx];

        if (c1 == c2) {
            answer[w2Idx] = i;
            w2Idx++;
        } 
        else if (!changed && (w2Idx == m - 1 || (lastValidW1Idx[w2Idx + 1] != -1 && lastValidW1Idx[w2Idx + 1] > i))) {
            answer[w2Idx] = i;
            w2Idx++;
            changed = 1;
        }
    }

    if (w2Idx < m) {
        free(lastValidW1Idx);
        free(answer);
        *returnSize = 0;
        return NULL;
    }

    free(lastValidW1Idx);
    *returnSize = m;
    return answer;
}

File: 3302-find-the-lexicographically-smallest-valid-sequence/3302-find-the-lexicographically-smallest-valid-sequence.py
class Solution:
    def validSequence(self, word1: str, word2: str) -> list[int]:
        n = len(word1)
        m = len(word2)

        # Space Optimized: Allocated exactly bound to the length of word2
        lastValidW1Idx = [-1] * m

        j = m - 1
        for i in range(n - 1, -1, -1):
            # Direct string indices lookups bypass list allocation delays
            if j >= 0 and word1[i] == word2[j]:
                lastValidW1Idx[j] = i
                j -= 1

        answer = [0] * m
        w2Idx = 0
        changed = False

        for i in range(n):
            if w2Idx >= m:
                break
                
            c1 = word1[i]
            c2 = word2[w2Idx]

            # Case 1: Fast exact matching track
            if c1 == c2:
                answer[w2Idx] = i
                w2Idx += 1
            # Case 2: Skip/Mismatch configuration slot evaluation
            elif not changed and (w2Idx == m - 1 or (lastValidW1Idx[w2Idx + 1] != -1 and lastValidW1Idx[w2Idx + 1] > i)):
                answer[w2Idx] = i
                w2Idx += 1
                changed = True

        if w2Idx < m:
            return []

        return answer
